keps/sig-network/4974-deprecate-endpoints/README.md

```diff
@@ -0,0 +1,659 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The `v1.Endpoints` API has been essentially (though not _actually_)
+deprecated since EndpointSlices became GA in 1.21. Several new Service
+features (such as dual-stack and topology, not to mention "services
+with more than 1000 endpoints") are implemented only for
+EndpointSlice, not for Endpoints. Kube-proxy no longer uses Endpoints
+ever, for anything, and the Gateway API conformance tests also require
+implementations to use EndpointSlices, so Gateway implementations
+don't use Endpoints either.
+
+Despite this, kube-controller-manager still does all of the work of
+managing Endpoints objects for all Services, and a cluster cannot pass
+the conformance test suite unless the Endpoints and EndpointSlice
+Mirroring controllers are running, even though in many cases nothing
+will ever look at the output of the Endpoints controller (and the
+EndpointSlice Mirroring controller will never output anything).
+
+While Kubernetes's API guarantees make it essentially impossible to
+ever actually fully remove Endpoints, we should at least move toward a
+world where most users run Kubernetes with the Endpoints and
+EndpointSlice Mirroring controllers disabled.
+
+## Motivation
+
+### Goals
+
+- Officially declare v1.Endpoints to be deprecated. Update
+  documentation and put out appropriate communications (blog posts,
+  etc) to ensure that end users are aware of this deprecation.
+
+- Add warnings (e.g., `Warning:` headers on Endpoints create/update)
+  to alert users of the fact that Endpoints is deprecated.
+
+- Update the e2e test suite to make it possible to run it in a "no
+  Endpoints controller" configuration, by rewriting some tests and
+  adding feature tags to others.
+
+- Update the conformance test suite to not require the Endpoints and
+  EndpointSlice Mirroring controllers to be running, by rewriting some
+  tests and demoting others from conformance.
+
+- Explicitly document that disabling `endpoints-controller` and/or
+  `endpointslice-mirroring-controller` via kube-controller-manager's
+  `--controllers` flag is a supported and conforming configuration.
+
+- Come up with a plan to clean up stale Endpoints in clusters that
+  aren't running the Endpoints controller.
+
+```
+<<[UNRESOLVED kubernetes.default ]>>
+
+- MAYBE change kube-apiserver to optionally not generate Endpoints for
+  kubernetes.default, though this would require adding a
+  kube-apiserver configuration option, and the benefit from removing
+  just that 1 Endpoints object is small. Perhaps instead it could just
+  add an annotation to the object pointing out the fact that Endpoints
+  is deprecated.
+
+<<[/UNRESOLVED]>>
+```
+
+```
+<<[UNRESOLVED disable-by-default ]>>
+
+- MAYBE eventually move `endpoints-controller` and
```

Comment:
I agree with moving towards being able to turn off Endpoints controller by default. I don't see how we'd ever turn off EndpointsMirroring controller by default without it being a hard break of behavior tied to a v1 API.